go/chat/attachments/gallery.go

```diff
@@ -0,0 +1,116 @@
+package attachments
+
+import (
+	"github.com/keybase/client/go/chat/globals"
+	"github.com/keybase/client/go/chat/utils"
+	"github.com/keybase/client/go/protocol/chat1"
+	"github.com/keybase/client/go/protocol/gregor1"
+	"golang.org/x/net/context"
+)
+
+type NextMessageOptions struct {
+	BackInTime bool
+	ImagesOnly bool
+}
+
+type Gallery struct {
+	globals.Contextified
+	utils.DebugLabeler
+
+	PrevStride, NextStride int
+}
+
+func NewGallery(g *globals.Context) *Gallery {
+	return &Gallery{
+		Contextified: globals.NewContextified(g),
+		DebugLabeler: utils.NewDebugLabeler(g.GetLog(), "Attachments.Gallery", false),
+		NextStride:   10,
+		PrevStride:   50,
+	}
+}
+
+func (g *Gallery) NextMessage(ctx context.Context, uid gregor1.UID,
+	convID chat1.ConversationID, msgID chat1.MessageID, opts NextMessageOptions) (res *chat1.MessageUnboxed, err error) {
+	eligible := func(msg chat1.MessageUnboxed) bool {
+		if !msg.IsValid() {
+			return false
+		}
+		body := msg.Valid().MessageBody
+		typ, err := body.MessageType()
+		if err != nil || typ != chat1.MessageType_ATTACHMENT {
+			return false
+		}
+		if !opts.ImagesOnly {
+			return true
+		}
+		md := body.Attachment().Object.Metadata
+		atyp, err := md.AssetType()
+		if err != nil {
+			return false
+		}
+		return atyp == chat1.AssetMetadataType_IMAGE || atyp == chat1.AssetMetadataType_VIDEO
+	}
+	if opts.BackInTime {
```

Comment:
i think these blocks would be easier to read if `opts.BackInTime` just determined the pagination (maybe `Pull` could  take a `reversed` param so the loops could just iterate over the response?

Reply:
I can swing by to discuss this.